### PR TITLE
Adding initial version of cord-19 dataset

### PR DESCRIPTION
**Pull request by ggdupont**
Initial version only reading the metadata in CSV.

### Checklist:
- [x] Create the dataset script /datasets/my_dataset/my_dataset.py using the template
- [x] Fill the _DESCRIPTION and _CITATION variables
- [x] Implement _infos(), _split_generators() and _generate_examples()
- [x] Make sure that the BUILDER_CONFIGS class attribute is filled with the different configurations of the dataset and that the BUILDER_CONFIG_CLASS is specified if there is a custom config class.
- [x] Generate the metadata file dataset_infos.json for all configurations
- [x] Generate the dummy data dummy_data.zip files to have the dataset script tested and that they don't weigh too much (<50KB)
- [x] Add the dataset card README.md using the template and at least fill the tags
- [x] Both tests for the real data and the dummy data pass.

### TODO:
- [x] add more metadata
- [x] add full text
- [x] add pre-computed document embedding

### CODE REVIEW

File: datasets/cord-19/README.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+annotations_creators:
+- no-annotation
+language_creators:
+- expert-generated
+languages:
+- en
+licenses:
+- 'other-multiple licenses: ''cc0'', ''hybrid-oa'', ''els-covid'', ''no-cc'', ''cc-by-nc-sa'',
+  ''cc-by'', ''gold-oa'', ''biorxiv'', ''green-oa'', ''bronze-oa'', ''cc-by-nc'',
+  ''medrxiv'', ''cc-by-nd'', ''arxiv'', ''unk'', ''cc-by-sa'', ''cc-by-nc-nd'''
```
Comment (lhoestq) [inline datasets/cord-19/README.md:11]: Can you list the licenses like this
```
licenses:
- license-name-1
- license-name-2
- other-license-name-3-that-is-not-available-in-the-tagging-app
```

?

You can use the tagging app to help you do that.
More info here : https://github.com/huggingface/datasets/blob/master/ADD_NEW_DATASET.md#tag-the-dataset-and-write-the-dataset-card

File: datasets/cord-19/cord-19.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
+# coding=utf-8
```
Comment (lhoestq) [inline datasets/cord-19/cord-19.py:1]: Can you move the directory from `cord-19` to `cord19` ? We're using snake_case convention. Can you also rename for the dataset script file name accordingly ?

File: datasets/cord-19/cord-19.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                    "title": datasets.Value("string"),
+                    "doi": datasets.Value("string"),
+                    "abstract": datasets.Value("string"),
+                    "publish_time": datasets.Value("string"),
+                    "authors": datasets.Value("string"),
+                    "journal": datasets.Value("string"),
+                    "url": datasets.Value("string"),
+                    # TODO "full_text" (complex nested json)
+                    # TODO "bib_entries" (complex nested json)
+                    # TODO "doc_embeddings" (separated file to be link by paper_id?)
```
Comment (lhoestq) [inline datasets/cord-19/cord-19.py:89]: do you want to add them in this PR ?

File: datasets/cord-19/dataset_infos.json [hunk 4]
```diff
@@ -0,0 +1 @@
...tring", "id": null, "_type": "Value"}, "source_x": {"dtype": "string", "id": null, "_type": "Value"}, "title": {"dtype": "string", "id": null, "_type": "Value"}, "doi": {"dtype": "string", "id": null, "_type": "Value"}, "abstract": {"dtype": "string", "id": null, "_type": "Value"}, "publish_time": {"dtype": "string", "id": null, "_type": "Value"}, "authors": {"dtype": "string", "id": null, "_type": "Value"}, "journal": {"dtype": "string", "id": null, "_type": "Value"}, "url": {"dtype": "string", "id": null, "_type": "Value"}}, "post_processed": null, "supervised_keys": null, "builder_name": "cord19", "config_name": "cord-19", "version": {"version_str": "0.0.0", "description": null, "major": 0, "minor": 0, "patch": 0}, "splits": {"train": {"name": "train", "num_bytes": 496247275, "num_examples": 368618, "dataset_name": "cord19"}}, "download_checksums": {"https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/historical_releases/cord-19_2020-11-29.tar.gz": {"num_bytes": 6142360818, "checksum": "56df7c715beaf8b84435f91b27fd7c8d9d1f50c6d04804bcf490e541d19d1783"}}, "download_size": 6142360818, "post_processing_size": null, "dataset_size": 496247275, "size_in_bytes": 6638608093}}
```
Comment (lhoestq) [inline datasets/cord-19/dataset_infos.json:1]: after the directory name change you'll have to regenerate this file

**Review by lhoestq — COMMENTED** [review]
Awesome ! thanks for adding this one

I left a few comments

File: datasets/cord-19/cord-19.py (hunk 3 above)
Comment (ggdupont) [reply datasets/cord-19/cord-19.py:89]: it's in progress but I can do another PR if you feel it's better

Comment (lhoestq) [reply datasets/cord-19/cord-19.py]: It's ok to add them in this PR
We usually do one PR = one dataset

Comment (ggdupont) [reply datasets/cord-19/cord-19.py]: That's what I had in mind (adding more configurations with the extra data).
I might not finished before the end of the hackathon though (and the dataset is perfectly usable with only metadata)

Comment (lhoestq) [reply datasets/cord-19/cord-19.py]: Ok :) feel free to ping me when you update the dataset

Comment (ggdupont) [reply datasets/cord-19/cord-19.py]: ping ;-)

File: datasets/cord19/README.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  "other-els-covid",
+  "other-no-cc",
+  "other-gold-oa",
+  "other-green-oa",
+  "other-bronze-oa",
+  "other-biorxiv",
+  "other-arxiv",
+  "other-medrxiv",
+  "other-unk"
+]
```
Comment (lhoestq) [inline datasets/cord19/README.md:27]:
```suggestion
licenses:
- cc-by-nc-sa-2.0
- cc-by-nc-2.0
- cc-by-nd-2.0
- cc-by-sa-2.0
- cc-by-nc-nd-2.0
- cc-by-nc-1.0
- other-cc0
- other-hybrid-oa
- other-els-covid
- other-no-cc
- other-gold-oa
- other-green-oa
- other-bronze-oa
- other-biorxiv
- other-arxiv
- other-medrxiv
- other-unk
```

File: datasets/cord19/README.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+### Languages
+
+The dataset is in english (en).
+
+## Dataset Structure
+
+### Data Instances
+
+TODO
```
Comment (lhoestq) [inline datasets/cord19/README.md:89]: Could you add one example from the dataset here ?

You can get one by using `dataset["train"][0]` for example

File: datasets/cord19/README.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+- `pdf_json_files`:  A `List[str]`-valued field containing paths from the root of the current data dump version to the parses of the paper PDFs into JSON format.  Multiple paths are semicolon-separated.  Example: `document_parses/pdf_json/4eb6e165ee705e2ae2a24ed2d4e67da42831ff4a.json; document_parses/pdf_json/d4f0247db5e916c20eae3f6d772e8572eb828236.json`
+- `pmc_json_files`:  A `List[str]`-valued field.  Same as above, but corresponding to the full text XML files downloaded from PMC, parsed into the same JSON format as above.
+- `url`: A `List[str]`-valued field containing all URLs associated with this paper.  Semicolon-separated.
+- `s2_id`:  A `str`-valued field containing the Semantic Scholar ID for this paper.  Can be used with the Semantic Scholar API (e.g. `s2_id=9445722` corresponds to `http://api.semanticscholar.org/corpusid:9445722`)
+
+Extra fields based on selected configuration during loading:
+
+- `fulltext`: A `str`-valued field containing the concatenation of all text sections from json (itself extracted from pdf)
+- `doc_embeddings`: A `str`-valued field containing document embeddings as a vector of floats separated by ','.
```
Comment (lhoestq) [inline datasets/cord19/README.md:118]: Can you explain how the embeddings were computed ?

File: datasets/cord19/cord19.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            if not os.path.isdir(fulltext_dir_path):
+                shutil.unpack_archive(os.path.join(data_dir, "2020-11-29/document_parses.tar.gz"), fulltext_dir_path)
+            files["fulltext"] = fulltext_dir_path
+
+        if "embeddings" in self.config.name:
+            embeddings_dir_path = os.path.join(data_dir, "2020-11-29/cord_19_embeddings")
+            if not os.path.isdir(embeddings_dir_path):
+                shutil.unpack_archive(
+                    os.path.join(data_dir, "2020-11-29/cord_19_embeddings.tar.gz"), embeddings_dir_path
+                )
```
Comment (lhoestq) [inline datasets/cord19/cord19.py:132]: Can you use `dl_manager.extract` instead ?

File: datasets/cord19/cord19.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+            "url": datasets.Value("string"),
+        }
+
+        if "fulltext" in self.config.name:
+            # adding full_text
+            features_dict["fulltext"] = datasets.Value("string")
+
+        if "embeddings" in self.config.name:
+            # adding embeddings
+            features_dict["doc_embeddings"] = datasets.Sequence(datasets.Value("float64"))
```
Comment (lhoestq) [inline datasets/cord19/cord19.py:101]: Should it be float64 or float32 ?

**Review by lhoestq — COMMENTED** [review]
That's really cool ! Thanks for adding the full texts and the embeddings

Good job :) 

I left a few comments

Can you also remove all the `dummy_data` folders and their content ? We only keep the `dummy_data.zip files`.

Moreover it looks like the dummy_data.zip files are quite big (90KB), could you try to reduce their sizes ? (ideally around 10-20KB). To do so feel free to remove all the unnecessary files and chunks of texts. The idea is to only keep a few examples with only a few sentences

File: datasets/cord19/README.md (hunk 6 above)
Comment (ggdupont) [reply datasets/cord19/README.md:87]: sure

File: datasets/cord19/cord19.py (hunk 9 above)
Comment (ggdupont) [reply datasets/cord19/cord19.py:101]: The values seems to have a high precision and necessitate float64
```
ug7v899j,-2.939983606338501,-6.312200546264648,-1.0459030866622925,5.164162635803223 [...]
```

Comment (lhoestq) [reply datasets/cord19/cord19.py:101]: Ok :)

File: datasets/cord19/cord19.py (hunk 8 above)
Comment (ggdupont) [reply datasets/cord19/cord19.py:132]: Tested locally: there is only 2 files to untar and it seems much slower than with shutil (roughly 5 times slower).

File: datasets/cord19/README.md (hunk 7 above)
Comment (ggdupont) [reply datasets/cord19/README.md:116]: I'll add a link to the paper: https://arxiv.org/abs/2004.07180

File: datasets/cord19/cord19.py (hunk 8 above)
Comment (lhoestq) [reply datasets/cord19/cord19.py:132]: Interesting ! Maybe we can use shutil to improve the speed of our file extraction.

Can you still change for dl_manager.extract ? The dl_manager is responsible for handling cache files.

**Comment by lhoestq** [comment]
Hi @ggdupont !
Have you had a chance to take a look at my suggestions ?
Feel free to ping me if you have questions or when you're ready for a review

**Comment by ggdupont** [comment]
> Hi @ggdupont !
> Have you had a chance to take a look at my suggestions ?
> Feel free to ping me if you have questions or when you're ready for a review

Yes I did, just busy period (and no time on weekend right now ;-) )

**Comment by ggdupont** [comment]
With some delay, reduced the dummy data and had t rebase

**Comment by lhoestq** [comment]
Thanks !

It looks like the rebase messed up the github diff for this PR (2.000+ files changed)
Could you create another branch and another PR please ?

**Comment by ggdupont** [comment]
Cleaned PR: https://github.com/huggingface/datasets/pull/1850

**Pull request closed without merging**